docs/current/guides/snippets/build-test-publish-php/dagger.php

```diff
@@ -0,0 +1,277 @@
+<?php
+// include auto-loader
+include 'vendor/autoload.php';
+
+use GraphQL\Client;
+
+class DaggerPipeline {
+
+  private $client;
+
+  // constructor
+  public function __construct() {
+    // initialize client with
+    // endpoint from environment
+    $sessionPort = getenv('DAGGER_SESSION_PORT') or throw new Exception("DAGGER_SESSION_PORT doesn't exist");
+    $sessionToken = getenv('DAGGER_SESSION_TOKEN') or throw new Exception("DAGGER_SESSION_TOKEN doesn't exist");
+    $this->client = new Client(
+      'http://127.0.0.1:' . $sessionPort . '/query',
+      ['Authorization' => 'Basic ' . base64_encode($sessionToken . ':')]
+    );
+  }
+
+  // build base image
+  public function buildBaseImage() {
+    // get host working directory
+    $sourceQuery = <<<QUERY
+    query {
+      host {
+        directory (path: ".", exclude: ["vendor", "ci"]) {
+          id
+        }
+      }
+    }
+    QUERY;
+    $sourceDir = $this->executeQuery($sourceQuery);
+
+    // build runtime image
+    // install tools and PHP extensions
+    // configure Apache webserver root and rewriting
+    $runtimeQuery = <<<QUERY
+    query {
+      container {
+        from(address: "php:8.2-apache-buster") {
+          withExec(args: ["apt-get", "update"]) {
+            withExec(args: ["apt-get", "install", "--yes", "git-core"]) {
+              withExec(args: ["apt-get", "install", "--yes", "zip"]) {
+                withExec(args: ["docker-php-ext-install", "pdo", "pdo_mysql", "mysqli"]) {
+                  withExec(args: ["sh", "-c", "sed -ri -e 's!/var/www/html!/var/www/public!g' /etc/apache2/sites-available/*.conf"]) {
+                    withExec(args: ["sh", "-c", "sed -ri -e 's!/var/www/!/var/www/public!g' /etc/apache2/apache2.conf /etc/apache2/conf-available/*.conf"]) {
+                      withExec(args: ["a2enmod", "rewrite"]) {
+                        id
+                      }
+                    }
+                  }
+                }
+              }
+            }
+          }
+        }
+      }
+    }
+    QUERY;
+    $runtime = $this->executeQuery($runtimeQuery);
+
+    // add application source code
+    // set file permissions
+    // set environment variables
+    $appQuery = <<<QUERY
```

Comment:
I would be tempted to move the app concern into a separate function. Notice how a more descriptive name - `buildRuntimeImage` - is forcing us to do the right thing here and separate concerns. I would extract the app query & `composer install` command into `buildAppImage` function.

Reply:
Done